infrahub_sdk/store.py

```diff
@@ -20,130 +23,336 @@ def get_schema_name(schema: str | type[SchemaType] | None = None) -> str | None:
-        hfid = node.get_human_friendly_id_as_string(include_kind=True)
+    def count(self) -> int:
+        return len(self._objs)
 
-        if not key and not hfid:
-            raise ValueError("Cannot store node without human friendly ID or key.")
+    def set(self, node: InfrahubNode | InfrahubNodeSync | CoreNode | CoreNodeSync, key: str | None = None) -> None:
+        self._objs[node._internal_id] = node
 
         if key:
-            node_kind = node._schema.kind
-            self._store[node_kind][key] = node
+            self._keys[key] = node._internal_id
 
-        if hfid:
-            self._store_by_hfid[hfid] = node
+        if node.id:
+            self._uuids[node.id] = node._internal_id
+
+        if hfid := node.get_human_friendly_id():
+            self._hfids[node.get_kind()][tuple(hfid)] = node._internal_id
+
+    def get(  # type: ignore[no-untyped-def]
+        self,
+        key: str,
+        kind: type[SchemaType | SchemaTypeSync] | str | None = None,
+        raise_when_missing: bool = True,
+    ) -> InfrahubNode | InfrahubNodeSync | CoreNode | CoreNodeSync | None:
+        found_invalid = False
 
-    def _get(self, key: str, kind: str | type[SchemaType] | None = None, raise_when_missing: bool = True):  # type: ignore[no-untyped-def]
         kind_name = get_schema_name(schema=kind)
-        if kind_name and kind_name not in self._store and key not in self._store[kind_name]:  # type: ignore[attr-defined]
-            if not raise_when_missing:
-                return None
-            raise NodeNotFoundError(
-                node_type=kind_name,
-                identifier={"key": [key]},
-                message="Unable to find the node in the Store",
-            )
 
-        if kind_name and kind_name in self._store and key in self._store[kind_name]:  # type: ignore[attr-defined]
-            return self._store[kind_name][key]  # type: ignore[attr-defined]
+        try:
+            return self._get_by_internal_id(key, kind=kind_name)
+        except NodeInvalidError:
+            found_invalid = True
+        except NodeNotFoundError:
+            pass
+
+        try:
+            return self._get_by_id(key, kind=kind_name)
+        except NodeInvalidError:
+            found_invalid = True
+        except NodeNotFoundError:
+            pass
 
-        for item in self._store.values():  # type: ignore[attr-defined]
-            if key in item:
-                return item[key]
+        try:
+            return self._get_by_key(key, kind=kind_name)
+        except NodeInvalidError:
+            found_invalid = True
+        except NodeNotFoundError:
+            pass
+
+        try:
+            return self._get_by_hfid(key, kind=kind_name)
+        except NodeNotFoundError:
+            pass
 
         if not raise_when_missing:
             return None
+
+        if kind and found_invalid:
+            raise NodeInvalidError(
```

Comment:
This error is raised when running `infrahub_edge.py` would be good with a pipeline in the Infrahub repo for this PR.